File: OO_design_patterns_training_2022-11/Solutions/C#/Exercise/Decorators.cs
using System;
using System.Collections.Generic;

namespace Exercise
{
    public abstract class Component
    {
        public abstract List<string> GetData();
    }

    public class ConcreteComponent : Component
    {
        private readonly List<string> m_data;

        public ConcreteComponent(List<string> data) { m_data = data; }
        public override List<string> GetData() { return m_data; }
    }

    public class Decorator : Component
    {
        private readonly Component m_component;

        public Decorator(Component component) { m_component = component; }
        public override List<string> GetData() { return m_component.GetData(); }
    }

    public class OrderDecorator : Decorator
    {
        public OrderDecorator(Component component) : base(component) {}

        public override List<string> GetData()
        {
            var data = base.GetData();
            data.Sort();
            return data;
        }
    }

    public class ReverseDecorator : Decorator
    {
        public ReverseDecorator(Component component) : base(component) {}

        public override List<string> GetData()
        {
            var data = base.GetData();

            for (var i = 0; i < data.Count; i++)
            {
                data[i] = ReverseText(data[i]);
            }

            return data;
        }

        private static string ReverseText(string text)
        {
            char[] array = text.ToCharArray();
            Array.Reverse(array);
            return new string(array);
        }
    }

    public class FilterDecorator : Decorator
    {
        public FilterDecorator(Component component) : base(component) {}

        public override List<string> GetData()
        {
            var data = base.GetData();
            data.RemoveAll(d => d.Contains("es"));
            return data;
        }
    }
}

File: OO_design_patterns_training_2022-11/MySolutions/C++/Exercise/src/Decorators.cpp
#include "Decorators.h"

#include <algorithm>

std::vector<std::string> ReverseDecorator::GetData() const
{
    auto data = Decorator::GetData();
    for (size_t i = 0; i < data.size(); i++)
    {
        data[i] = ReverseText(data[i]);
    }
    return data;
}

std::string ReverseDecorator::ReverseText(const std::string& text)
{
    return std::string{text.rbegin(), text.rend()};
}

std::vector<std::string> OrderDecorator::GetData() const
{
    auto data = Decorator::GetData();
    std::sort(data.begin(), data.end());
    return data; 
}

std::vector<std::string> FilterDecorator::GetData() const
{
    auto data = Decorator::GetData();
    data.erase(std::remove_if(data.begin(), data.end(), [](const std::string& str){ return str.find("es") != std::string::npos; }), data.end());
    return data;
}
